packages/shared-ini-file-loader/src/slurpFile.ts

```diff
@@ -0,0 +1,63 @@
+// ToDo: Change to "fs/promises" when supporting nodejs>=14
+import { promises as fsPromises } from "fs";
+
+const { readFile, stat } = fsPromises;
+
+type callbacks = { resolve: Function; reject: Function };
+type fileStatus = {
+  contents: string;
+  isReading: boolean;
+  lastModified: Date;
+  requestQueue: callbacks[];
+};
+
+const fileReadHash: { [key: string]: fileStatus } = {};
+
+const callReadFile = (path: string, cbs: callbacks) => {
+  fileReadHash[path].requestQueue.push(cbs);
+  readFile(path, "utf8")
+    .then((data) => {
+      // File read successful
+      fileReadHash[path].contents = data;
+      fileReadHash[path].requestQueue.forEach(({ resolve }) => resolve(data));
+    })
+    .catch((err) => {
+      fileReadHash[path].requestQueue.forEach(({ reject }) => reject(err));
+    })
+    .finally(() => {
+      fileReadHash[path].isReading = false;
+      fileReadHash[path].requestQueue = [];
+      // Update last modified.
+      stat(path).then(({ mtime }) => {
+        fileReadHash[path].lastModified = mtime;
+      });
+    });
+};
+
+export const slurpFile = (path: string) =>
+  new Promise<string>((resolve, reject) => {
+    if (!fileReadHash[path]) {
+      // File not read yet, set file isReading to true.
+      fileReadHash[path] = { isReading: true, lastModified: new Date(0), contents: "", requestQueue: [] };
+      // Read file.
+      callReadFile(path, { resolve, reject });
+    } else if (fileReadHash[path].isReading) {
+      // File currently being read. Add to request queue.
+      fileReadHash[path].requestQueue.push({ resolve, reject });
+    } else {
+      // File already read, or read failed.
+      if (fileReadHash[path].lastModified === null) {
+        // Read failed. Read file again.
+        callReadFile(path, { resolve, reject });
+      } else {
+        // Read file only if it's modified.
+        stat(path).then(({ mtime }) => {
```

Comment:
So I assume stat will not increase the EMFILE count?

Reply:
No. From Node.js docs on [stats.mtime](https://nodejs.org/api/fs.html#statsmtime)
> The timestamp indicating the last time this file was modified.

Reply:
> So I assume stat will not increase the EMFILE count?

Stat will increase EMFILE count by one when the operation is in progress, but the number of files open at any point in time will reduce as:
* We're no longer reading file at `path` in parallel.
* We're not reading file if it's not modified since the last time it was read.
* The functions `stat` and `readFile` are not called in parallel for a particular path, and they close the file descriptor once operation is complete.

Reply:
>and they close the file descriptor once operation is complete

I hope so, the same hope was with the current behaviour...

at least with this change the config and credential file would at max increment the EMFILE count by 4 during a lambda invocation, because queued